File: IOI/ioi2002/day0/PracticeAuxiliary/string/source/oracle.c
#include "oracle.h"
#include <stdio.h>
#include <string.h>

#define INFILE "string.in"
#define OUTFILE "string.out"
#define LOGFILE "string.log"

static FILE *fin;
static FILE *fout;
static char sin[255];
static int oracle_cnt=-99;

static void ABOUT(char *erm)
{
	printf("%s\n",erm); 	
	exit(2);
}

void start_string(void)
{
	int cc=0;
	if (oracle_cnt!=-99) ABOUT("Error: You called start_string() 2 times.");    	
	
	oracle_cnt=0;
	fin=fopen(INFILE,"r");
	if (!fin) ABOUT("Error: Cannot open 'string.in'.");    	
	fout=fopen(OUTFILE,"w");
	while(!feof(fin)) {	
		fscanf(fin,"%c",&sin[cc]);
		if (!((sin[cc]==10) || (sin[cc]==13))) {
			// to search ERROR input string.
			// if ((sin[cc]!='A') && (sin[cc]!='T') && (sin[cc]!='G') && (sin[cc]!='C'))
			//	ABOUT("ERROR: other character except A,T,G,C appeared.");
			cc++;
		}
	}
	sin[cc]=0;
}

int oracle_call(char *s)
{
	int i;
	char *pch;	
	for(i=0;i<strlen(s);i++)
		if ((s[i]!='A') && (s[i]!='T') && (s[i]!='G') && (s[i]!='C'))
			ABOUT("ERROR: other character except A,T,G,C appeared.");		
	if (oracle_cnt==-99) ABOUT("ERROR: start_string() called at first.\n");	
	if (strlen(s)==0) ABOUT("ERROR: S should not be null.");
	if (strlen(s)>=255) ABOUT("ERROR: The length of query S shold be less than 255 chars.");
	oracle_cnt++;
	pch = strstr (sin,s);
	if (pch==NULL) return 0; else return 1;		
	return 0;
}

void answer_string (char *s)
{
	if (oracle_cnt==-99) ABOUT("ERROR: start_string() called at first\n");			 
	fprintf(fout,"%d\n",oracle_cnt);
	fprintf(fout,"%s\n",s);
	fclose(fout);
	fclose(fin);
    exit(0);
}

File: IOI/ioi2002/day0/PracticeAuxiliary/string/source/oracle.pas
unit oracle;

interface

procedure start_string;
function  oracle_call(s:String) : integer;
procedure answer_string(s:String);

implementation

Uses SysUtils;

var
	oracle_cnt:integer;
	fin,fout:text;
	sin:String;

procedure ABOUT(e:String);
begin
	writeln (e);
	halt (2);
end;

procedure start_string;
begin
	oracle_cnt:=0;
	assign (fin, 'string.in' );	
	{$I-}	
	reset(fin);
	{$I+}
	if (IOresult <> 0) then ABOUT('Error: Cannot open ''string.in''.');
	readln(fin,sin);
	assign (fout, 'string.out' );	
	{$I-}		
	rewrite(fout);	
	{$I+}	
	if (IOresult <> 0) then ABOUT('Error: Cannot open ''string.out''.');	
end;

function oracle_call(s:String) : integer ;
var i:integer;
    tt:integer;
begin
	if oracle_cnt=-99 then ABOUT('ERROR: start_string() called at first.');
	for i:=1 to length(s) do
		if ((s[i]<>'A') and (s[i]<>'G') and (s[i]<>'T') and (s[i]<>'C')) then
		ABOUT('ERROR: other character except A,T,G,C appeared.');
	if (length(s)=0) then ABOUT('ERROR: S should not be null.');
	if (length(s)>255) then ABOUT('ERROR: The length of query S shold be less than 255 chars.');
	inc(oracle_cnt);
	if (pos(sin, s) <> 0) then tt:=1 else tt:=0;	
	oracle_call:=tt;
end;


procedure answer_string(s:String);
begin
	if oracle_cnt=-99 then ABOUT('ERROR: start_string() called at first.');
	writeln(fout,oracle_cnt);
	writeln(fout,s);	
	close(fout);
	close(fin);	
	halt;
end;

begin
	oracle_cnt:=-99;
end.


